vertx-web/src/main/java/io/vertx/ext/web/impl/HttpServerRequestWrapper.java

```diff
@@ -149,6 +152,37 @@ public String query() {
     return query;
   }
 
+  @Override
+  public MultiMap params() {
+    if (!modified) {
+      return delegate.params();
+    }
+    if (params == null) {
+      params = MultiMap.caseInsensitiveMultiMap();
+      // if there is no query it's not really needed to parse it
+      if (query != null) {
+        QueryStringDecoder queryStringDecoder = new QueryStringDecoder(uri);
```

Comment:
@pmlopes, I just realized that one side-effect of caching the params would be that `reroute()` now works with query params. Here's a test that would fail before this PR.
```java
@Test
public void testRerouteWithQueryParams() throws Exception {
    router.get("/other").handler(ctx -> {
      assertEquals("value2", ctx.request().params().get("param1"));
      ctx.response().end("/other");
    });
    router.get("/base").handler(ctx -> ctx.reroute("/other?param1=value2"));

    testRequest(HttpMethod.GET, "/base?param1=value1", 200, "OK", "/other");
  }
```

Just wanted to point this out and that an update to the docs might be needed if we keep this. [Here's the part](https://vertx.io/docs/vertx-web/java/#_reroute) where it's stating reroute only works on paths.